modyn/trainer_server/internal/grpc/trainer_server_grpc_servicer.py

```diff
@@ -222,6 +247,83 @@ def get_training_status(
         cleaned_kwargs = {k: v for k, v in response_kwargs_finished.items() if v is not None}
         return TrainingStatusResponse(**cleaned_kwargs)  # type: ignore[arg-type]
 
+    def get_values_from_queues(
+        self, training_id: int
+    ) -> Tuple[Optional[bool], Optional[int], Optional[int], Optional[int], Optional[int]]:
+        was_training = self._training_process_dict[training_id].was_training
+
+        if was_training:
+            (
+                is_training,
+                downsampling_num_batches,
+                downsampling_num_samples,
+                training_num_batches,
+                training_num_samples,
+            ) = self._handle_was_training(training_id)
+        else:
+            (
+                is_training,
+                downsampling_num_batches,
+                downsampling_num_samples,
+                training_num_batches,
+                training_num_samples,
+            ) = self._handle_was_not_training(training_id)
+
+        if is_training is not None:
+            self._training_process_dict[training_id].was_training = is_training
+
+        return (
+            is_training,
+            downsampling_num_batches,
+            downsampling_num_samples,
+            training_num_batches,
+            training_num_samples,
+        )
+
+    def _handle_was_not_training(
+        self, training_id: int
+    ) -> Tuple[Optional[bool], Optional[int], Optional[int], Optional[int], Optional[int]]:
+        downsampling_num_batches, downsampling_num_samples, is_training = self.get_status_downsampling(
+            training_id, timeout=15
+        )
+        if is_training:
+            # clean the downsampling queue (downsampling is ended)
+            self.clean_downsampling_queue(training_id)
+        # read the second queue only if the first one is empty
+        if downsampling_num_batches is None:
+            training_num_batches, training_num_samples, is_training = self.get_status_training(training_id, timeout=15)
+        else:
+            training_num_batches, training_num_samples = None, None
+        return (
+            is_training,
+            downsampling_num_batches,
+            downsampling_num_samples,
+            training_num_batches,
+            training_num_samples,
+        )
+
+    def _handle_was_training(
+        self, training_id: int
+    ) -> Tuple[Optional[bool], Optional[int], Optional[int], Optional[int], Optional[int]]:
+        training_num_batches, training_num_samples, is_training = self.get_status_training(training_id, timeout=15)
+        if not is_training:
+            # clean the training queue (epoch is finished)
+            self.clean_training_queue(training_id)
```

Comment:
Just for my understanding: Why did this become necessary?

Reply:
For a scenario like this

Read downsampling queue (long timeout) 
Queue empty, read from training queue (small timeout)
--- while waiting, the response is added to the downsampling queue ---
Get a valid value from the training queue => downsampling is finished, switch to the other bar. Move to training phase

Here we want to delete the response in the downsampling queue. Otherwise, the following might happen

Read training queue. Training is taking longer and no update is posted. So, switch to the downsampling queue
There you find the old message which would reopen the downsampling phase and close the training one

Reply:
Thank you for explaining.